cni/pkg/mount/mount.go

```diff
@@ -0,0 +1,52 @@
+package mount
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+)
+
+// RunSandboxed is an opinionated function to run a command in a lightweight sandbox.
+// This overrides files that are not desired in iptables execution (which runs on the node) by bind mounting them
+func RunSandboxed(lockFile string, args []string) error {
+	// In CNI, we are running the pod network namespace, but the host filesystem. Locking the host is both useless and harmful,
+	// as it opens the risk of lock contention with other node actors (such as kube-proxy), and isn't actually needed at all.
+	// Older iptables cannot turn off the lock explicitly, so we hack around it...
+	// Overwrite the lock file with the network namespace file (which is assumed to be unique).
+	// We are setting the lockfile to `r.NetworkNamespace`.
+	// /dev/null looks like a good option, but actually doesn't work (it will ensure only one actor can access it)
```

Comment:
The netns should work - would a simple `createTempfile/defer cleanTempfile` also work without asking the caller to provide their own lockfile and maintaining the internal convention of using `netns-as-junk-lock`?

Using the netns makes it harder for a process in the same namespace to abuse this I guess. If we stick with the namespace, can we simply get the current thread's netns path and use it from inside this func, rather than expecting someone to pass it in?

Reply:
The intent of using the network namespace was to avoid cases where the host filesystem is not writable. We _know_ the netns is there and its extrmely unlike to ahve something hold a lock against it.

We could get the current netns path but seems unnecessarily complex. This is an internal interface, it doesn't matter how hard it is to use. No human should ever call this.

Reply:
Right, but considering that inpod is adding another iptables impl and ideally we would like a single iptables helper we could use *everywhere* and not just from/with `istio-cni` it seems smarter to keep the iptables workarounds in the iptables helper lib so it can be used across different envs without `istio-cni`-specific concerns leaking.

Unshare is needed for the iptables wrapper. It really has nothing to do with `istio-cni`, other than the fact that `istio-cni` is currently what calls iptables - but that's about to change and I'd rather have one value-add `iptables` wrapper we can use in multiple spots.

Reply:
I don't really understand what the suggestion is? Make a new binary instead of a subcommand?

Reply:
Sorry - I'm suggesting 

- move `mount.go` under `tools/istio-iptables`
- move `runUnshare()`  to `tools/istio-iptables/pkg/cmd/root.go` as a new method to exec `iptables`
- keep `istio-cni unshare` as an optional flag, but rather than setting up the netns itself, it just invokes the iptables lib with a given netns path.

We should be able to do the mount/exec from there, no? 

So that any caller of the iptables helper lib can get the lockfile workaround without having to go thru the `istio-cni` binary - the `istio-cni` binary can/should still invoke iptables thru this wrapper internally tho.

Reply:
The goal overall sounds nice I just am not sure I 100% understand it. I don't mind moving functions whereever, that is just trivial re-organization. But in CNI, the `istio-iptables` binary which `tools/istio-iptables/pkg/cmd/root.go` compiles into is not used or present on the node. So we could move the function there (simple code refactor) or _add_ it to _also_ be there (even though we would never use it, at this point), but we cannot .

One point that may be important here is this needs to be a separate `exec`ed command so we can run in a new mount namespace, so its not just a simple function.
At least in the current approach - its plausible we could do the new mount namespace in process but I think it gets funky with goroutines, threads, etc

Reply:
> The goal overall sounds nice I just am not sure I 100% understand it. I don't mind moving functions whereever, that is just trivial re-organization. But in CNI, the `istio-iptables` binary which `tools/istio-iptables/pkg/cmd/root.go` compiles into is not used or present on the node. So we could move the function there (simple code refactor) or _add_ it to _also_ be there (even though we would never use it, at this point), but we cannot .
> 
> One point that may be important here is this needs to be a separate `exec`ed command so we can run in a new mount namespace, so its not just a simple function. At least in the current approach - its plausible we could do the new mount namespace in process but I think it gets funky with goroutines, threads, etc

Ah my bad then, I thought `istio-iptables` was also consumed as a lib by `istio-cni`, so we wouldn't need to exec a separate binary.


Nevermind this then, we can tackle it later if/when we do consolidate the iptables libraries/wrappers/etc.

Reply:
istio-cni consumes istio-iptables as a lib (cmd.ProgramIptables). But its needs to Exec out to the mounting logic so we can run it with `c.SysProcAttr = &syscall.SysProcAttr{Unshareflags: unix.CLONE_NEWNS}`